lib/wasi/src/runtime/resolver/inputs.rs

```diff
@@ -0,0 +1,254 @@
+
+/// A reference to *some* package somewhere that the user wants to run.
+///
+/// # Security Considerations
+///
+/// The [`PackageSpecifier::Path`] variant doesn't specify which filesystem a
+/// [`Source`][source] will eventually query. Consumers of [`PackageSpecifier`]
+/// should be wary of sandbox escapes.
+///
+/// [source]: crate::runtime::resolver::Source
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum PackageSpecifier {
+    Registry {
+        full_name: String,
+        version: VersionReq,
+    },
+    Url(Url),
+    /// A `*.webc` file on disk.
+    Path(PathBuf),
+}
+
+impl PackageSpecifier {
+    pub fn parse(s: &str) -> Result<Self, anyhow::Error> {
+        s.parse()
+    }
+}
+
+impl FromStr for PackageSpecifier {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // TODO: Replace this with something more rigorous that can also handle
+        // the locator field
+        let (full_name, version) = match s.split_once('@') {
+            Some((n, v)) => (n, v),
+            None => (s, "*"),
+        };
+
+        let invalid_character = full_name
+            .char_indices()
+            .find(|(_, c)| !matches!(c, 'a'..='z' | 'A'..='Z' | '0'..='9' | '.'| '-'|'_' | '/'));
+        if let Some((index, c)) = invalid_character {
+            anyhow::bail!("Invalid character, {c:?}, at offset {index}");
+        }
+
+        let version = version
+            .parse()
+            .with_context(|| format!("Invalid version number, \"{version}\""))?;
+
+        Ok(PackageSpecifier::Registry {
+            full_name: full_name.to_string(),
+            version,
+        })
+    }
+}
+
+impl Display for PackageSpecifier {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PackageSpecifier::Registry { full_name, version } => write!(f, "{full_name}@{version}"),
+            PackageSpecifier::Url(url) => Display::fmt(url, f),
+            PackageSpecifier::Path(path) => write!(f, "{}", path.display()),
+        }
+    }
+}
+
+/// A dependency constraint.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Dependency {
+    pub alias: String,
+    pub pkg: PackageSpecifier,
+}
+
+impl Dependency {
+    pub fn package_name(&self) -> Option<&str> {
+        match &self.pkg {
+            PackageSpecifier::Registry { full_name, .. } => Some(full_name),
+            _ => None,
+        }
+    }
+
+    pub fn alias(&self) -> &str {
+        &self.alias
+    }
+
+    pub fn version(&self) -> Option<&VersionReq> {
+        match &self.pkg {
+            PackageSpecifier::Registry { version, .. } => Some(version),
+            _ => None,
+        }
+    }
+}
+
+/// Some metadata a [`Source`][source] can provide about a package without
+/// needing to download the entire `*.webc` file.
+///
+/// [source]: crate::runtime::resolver::Source
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Summary {
```

Comment:
```suggestion
pub struct PackageSummary {
```

Already saw it in Deploy: `Summary` is confusing.

`PkgSummary` works too...

Reply:
I really don't want to go down this path because we'll have half a dozen different concepts all called `PackageXxx` (`PackageLoader`, `PackageResolver`, `PackageSummary`, `PackageInfo`, `PackageSource`).

This sounds like the [Smurf Naming Convention](https://blog.codinghorror.com/new-programming-jargon/#21):

> When almost every class has the same prefix. IE, when a user clicks on the button, a SmurfAccountView passes a SmurfAccountDTO to the SmurfAccountController. The SmurfID is used to fetch a SmurfOrderHistory which is passed to the SmurfHistoryMatch before forwarding to either SmurfHistoryReviewView or SmurfHistoryReportingView. If a SmurfErrorEvent occurs it is logged by SmurfErrorLogger to ${app}/smurf/log/smurf/smurflog.log

Reply:
Nature of the beast.

You took the naming from `cargo`.

For cargo probably half the code is just about dealing with packages, and it's rarely used outside of cargo.

For us these types are often used in external code , so then the naming is very suboptimal.

Reply:
I agree that `PackageSummary` is clearer than `Summary`. I actually thought that `Summary` was not the ideal name even before seeing @theduke suggestions.

Maybe we can use other name, but for sure `Summary` is too generic and doesn't say anything about what the type is about